File: zig/src/token.zig
const std = @import("std");

pub const ILLEGAL = "illegal";
pub const EOF = "eof";
pub const IDENT = "ident";
pub const INT = "int";

// operators
pub const ASSIGN = "=";
pub const PLUS = "+";
pub const MINUS = "-";
pub const BANG = "!";
pub const ASTERISK = "*";
pub const SLASH = "/";
pub const LT = "<";
pub const GT = ">";

pub const EQ = "==";
pub const NOT_EQ = "!=";

pub const COMMA = ",";
pub const SEMICOLON = ";";
pub const LPAREN = "(";
pub const RPAREN = ")";
pub const LBRACE = "{";
pub const RBRACE = "}";

pub const FUNCTION = "function";
pub const LET = "let";
pub const TRUE = "true";
pub const FALSE = "false";
pub const IF = "if";
pub const ELSE = "else";
pub const RETURN = "return";

pub const Token = struct {
    type: []const u8,
    literal: []const u8,
    pub fn format(
        self: Token,
        comptime fmt: []const u8,
        options: std.fmt.FormatOptions,
        writer: anytype,
    ) !void {
        _ = fmt;
        _ = options;

        try writer.print("Token{{ .type = \"{s}\", .literal = \"{s}\" }}", .{ self.type, self.literal });
    }
};

const keywords = std.ComptimeStringMap([]const u8, .{
    &.{ "fn", FUNCTION },
    &.{ "let", LET },
    &.{ "true", TRUE },
    &.{ "false", FALSE },
    &.{ "if", IF },
    &.{ "else", ELSE },
    &.{ "return", RETURN },
});

pub fn lookupIdent(identifier: []const u8) []const u8 {
    if (keywords.get(identifier)) |tok| {
        return tok;
    } else {
        return IDENT;
    }
}

File: elixir/lib/monkey/token.ex
defmodule Monkey.Token do
  defstruct [:type, :literal]

  @type token_type :: String.t()

  @type t :: %__MODULE__{
          type: token_type(),
          literal: String.t()
        }

  @tokens %{
    illegal: "ILLEGAL",
    eof: "EOF",

    # identifiers
    ident: "IDENT",
    int: "INT",

    # operators
    assign: "=",
    plus: "+",
    minus: "-",
    bang: "!",
    asterisk: "*",
    slash: "/",
    lt: "<",
    gt: ">",
    eq: "==",
    not_eq: "!=",

    # delimiters
    comma: ",",
    semicolon: ";",
    lparen: "(",
    rparen: ")",
    lbrace: "{",
    rbrace: "}",

    # keywords
    function: "FUNCTION",
    let: "LET",
    true: "TRUE",
    false: "FALSE",
    if: "IF",
    else: "ELSE",
    return: "RETURN",
    string: "STRING"
  }

  @keywords %{
    "fn" => @tokens.function,
    "let" => @tokens.let,
    "true" => @tokens.true,
    "false" => @tokens.false,
    "if" => @tokens.if,
    "else" => @tokens.else,
    "return" => @tokens.return
  }

  def new(token_type, char \\ nil) do
    %__MODULE__{
      type: token_type,
      literal: char
    }
  end

  def tokens() do
    @tokens
  end

  def tokens(key) do
    @tokens[key]
  end

  def lookup_identifier(identifier) do
    Map.get(@keywords, identifier, @tokens.ident)
  end
end
